### PR TITLE
Added preCreate, preEdit, preDelete, preShow, preList methods in the CRUDController

### PR DESCRIPTION
**Pull request by pulzarraider**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT

This PR adds some methdos to help develoipers overload the CRUDController´and reduce the code they have to duplicate in overloaded methods.

### CODE REVIEW

File: Resources/doc/reference/architecture.rst [hunk 1]
```diff
@@ -147,6 +147,18 @@ a set of automatically injected dependencies that are useful when implementing s
 scenarios. Refer to the existing ``CRUDController`` actions for examples of how to get
 the best out of them.
 
+In your overloaded CRUDContreller you can overload also these methods to limit
```
Comment (GromNaN) [inline Resources/doc/reference/architecture.rst]: There is a typo  on CRUDContreller

Comment (pulzarraider) [inline Resources/doc/reference/architecture.rst]: Fixed, thanks.

**Comment by Soullivaneuh** [comment]
@rande @pulzarraider This produce some errors like this on tests:

```
1) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testPreCreate
ReflectionException: Class  does not exist

/home/sullivan/projects/fork/SonataAdminBundle/Controller/CRUDController.php:657
/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:1742

2) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithModelManagerException with data set #0 ('', '')
ReflectionException: Class  does not exist

/home/sullivan/projects/fork/SonataAdminBundle/Controller/CRUDController.php:657
/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:2015

3) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithModelManagerException with data set #1 ('Foo', 'Foo')
ReflectionException: Class  does not exist

/home/sullivan/projects/fork/SonataAdminBundle/Controller/CRUDController.php:657
/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:2015

4) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithModelManagerException with data set #2 ('&lt;a href=&quot;http://foo&q.../a&gt;', '<a href="http://foo">Bar</a>')
ReflectionException: Class  does not exist

/home/sullivan/projects/fork/SonataAdminBundle/Controller/CRUDController.php:657
/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:2015

5) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithModelManagerException with data set #3 ('&lt;&gt;&amp;&quot;&#039;abcd...()[]\/', '<>&"'abcdefghijklmnopqrstuvwx...()[]\/')
ReflectionException: Class  does not exist
```

**Comment by rande** [comment]
@Soullivaneuh I fixed some of them from a previous PR : https://github.com/sonata-project/SonataAdminBundle/commit/9c4457918c809b340a6c9e2f109231bebaa9b720 and https://github.com/sonata-project/SonataAdminBundle/commit/8b80e43f7c9fbad66072ea41c97edbbcc635245b

**Comment by Soullivaneuh** [comment]
@rande sorry but seems not (`upstream` is your repo):

```bash
sullivan@sweethome:~/projects/fork/SonataAdminBundle(master)$ git fetch --all --prune
Fetching origin
Fetching upstream
Fetching nc
sullivan@sweethome:~/projects/fork/SonataAdminBundle(master)$ git merge upstream/master 
Already up-to-date.
sullivan@sweethome:~/projects/fork/SonataAdminBundle(master)$ phpunit 
PHPUnit 4.6.6 by Sebastian Bergmann and contributors.

Configuration read from /home/sullivan/projects/fork/SonataAdminBundle/phpunit.xml.dist

...............................................................  63 / 981 (  6%)
............................................................... 126 / 981 ( 12%)
............................................................... 189 / 981 ( 19%)
............................................................... 252 / 981 ( 25%)
............................................................... 315 / 981 ( 32%)
...FFFF..E..........EEEE....................................... 378 / 981 ( 38%)
............................................................... 441 / 981 ( 44%)
............................................................... 504 / 981 ( 51%)
............................................................... 567 / 981 ( 57%)
............................................................... 630 / 981 ( 64%)
............................................................... 693 / 981 ( 70%)
............................................................... 756 / 981 ( 77%)
............................................................... 819 / 981 ( 83%)
............................................................... 882 / 981 ( 89%)
............................................................... 945 / 981 ( 96%)
....................................

Time: 13.85 seconds, Memory: 83.25Mb

There were 5 errors:

1) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testPreCreate
ReflectionException: Class  does not exist

/home/sullivan/projects/fork/SonataAdminBundle/Controller/CRUDController.php:657
/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:1742

2) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithModelManagerException with data set #0 ('', '')
ReflectionException: Class  does not exist

/home/sullivan/projects/fork/SonataAdminBundle/Controller/CRUDController.php:657
/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:2015

3) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithModelManagerException with data set #1 ('Foo', 'Foo')
ReflectionException: Class  does not exist

/home/sullivan/projects/fork/SonataAdminBundle/Controller/CRUDController.php:657
/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:2015

4) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithModelManagerException with data set #2 ('&lt;a href=&quot;http://foo&q.../a&gt;', '<a href="http://foo">Bar</a>')
ReflectionException: Class  does not exist

/home/sullivan/projects/fork/SonataAdminBundle/Controller/CRUDController.php:657
/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:2015

5) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithModelManagerException with data set #3 ('&lt;&gt;&amp;&quot;&#039;abcd...()[]\/', '<>&"'abcdefghijklmnopqrstuvwx...()[]\/')
ReflectionException: Class  does not exist

/home/sullivan/projects/fork/SonataAdminBundle/Controller/CRUDController.php:657
/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:2015

--

There were 4 failures:

1) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testEditActionWithModelManagerException with data set #0 ('', '')
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    'sonata_flash_error' => Array (...)
 )

/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:1653

2) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testEditActionWithModelManagerException with data set #1 ('Foo', 'Foo')
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    'sonata_flash_error' => Array (...)
 )

/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:1653

3) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testEditActionWithModelManagerException with data set #2 ('&lt;a href=&quot;http://foo&q.../a&gt;', '<a href="http://foo">Bar</a>')
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    'sonata_flash_error' => Array (...)
 )

/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:1653

4) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testEditActionWithModelManagerException with data set #3 ('&lt;&gt;&amp;&quot;&#039;abcd...()[]\/', '<>&"'abcdefghijklmnopqrstuvwx...()[]\/')
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    'sonata_flash_error' => Array (...)
 )

/home/sullivan/projects/fork/SonataAdminBundle/Tests/Controller/CRUDControllerTest.php:1653

FAILURES!
Tests: 981, Assertions: 2807, Errors: 5, Failures: 4.
```

**Comment by pulzarraider** [comment]
Sorry, I missed this failing tests. Fixed in #2974.

**Comment by Koc** [comment]
@pulzarraider very useful, just want to say thanks

**Comment by pulzarraider** [comment]
Nice to hear that, thank you @Koc

**Comment by OskarStark** [comment]
Thank you @pulzarraider

**Comment by anacona16** [comment]
Any way to have postCreate, postEdit, etc?

Or how to redirect to a custom route after create an entity without overwrite the createAction?

**Comment by pulzarraider** [comment]
@anacona16 please create a new issue. This is closed PR.

**Pull request merged**